modules/datastore/src/DataDictionary/DateFormat/Parser.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace Drupal\datastore\DataDictionary\DateFormat;
+
+use Drupal\datastore\DataDictionary\GrammarInterface;
+use Drupal\datastore\DataDictionary\ParserInterface;
+use Drupal\datastore\DataDictionary\UnknownTokenException;
+
+/**
+ * Date format parser.
+ */
+class Parser implements ParserInterface {
+
+  /**
+   * Creates a date format parser.
+   *
+   * @param \Drupal\datastore\DataDictionary\GrammarInterface $grammar
+   *   Grammar for date format being parsed.
+   */
+  public function __construct(GrammarInterface $grammar) {
+    $this->grammar = $grammar;
+  }
+
+  /**
+   * Break up supplied date format string into lexemes.
+   *
+   * @param string $input_format
+   *   Date format string to lex.
+   *
+   * @return \Drupal\datastore\DataDictionary\FrictionlessDateFormatConverter\TokenInterface[]
+   *   Token list.
+   */
+  protected function lex(string $input_format): array {
```

Comment:
Even though this method has a high "cognitive complexity", I think that breaking it up into smaller methods would only decrease the readability of the `lex()` method. I've documented it as extensively as I know how, but this is one of those algorithms which is hard to break up, and difficult to understand the parts of in isolation. As such, I recommend this method be left composite.

Reply:
@clayliddell I agree, I think, about the cognitive complexity being acceptable here but I think we should think about isolating esoteric, self-contained/reusable things like parsers from our basic business logic. And if possible, avoid implementing and having to maintain them ourselves. Did you look into this library, which seems to provide strptime parsing for current PHP versions? 

https://packagist.org/packages/php81_bc/strftime

If this one is missing critical functionality I'd suggest we break off your date parser into its own library.